### PR TITLE
Make PipHistoricPerfData merge crashes more visible

### PR DESCRIPTION
**Pull request by Kr1sso**
Remove implicit casts and add some debugging capabilities for historic perf data crashes we are seeing rarely.

### CODE REVIEW

File: Public/Src/Engine/Processes/SandboxedProcessMac.cs [hunk 1]
```diff
@@ -414,7 +414,7 @@ internal override JobObject.AccountingInformation GetJobAccountingInfo()
                         ReadTransferCount = Convert.ToUInt64(m_perfAggregator.DiskBytesRead.Total),
                         WriteTransferCount = Convert.ToUInt64(m_perfAggregator.DiskBytesWritten.Total)
                     }),
-                    PeakMemoryUsage = (ulong)m_perfAggregator.PeakMemoryBytes.Maximum,
+                    PeakMemoryUsage = Convert.ToUInt64(m_perfAggregator.PeakMemoryBytes.Maximum),
```
Comment (Kr1sso) [inline Public/Src/Engine/Processes/SandboxedProcessMac.cs]: Potential culprit.

Comment (mpysson) [inline Public/Src/Engine/Processes/SandboxedProcessMac.cs]: >Convert.ToUInt64 [](start = 38, length = 16)

Convert.ToUInt64() still throws an OverflowException if PeakMemoryBytes is negative or too large. What does this change?

File: Public/Src/Engine/Scheduler/PipHistoricPerfData.cs [hunk 2]
```diff
@@ -176,23 +177,30 @@ public PipHistoricPerfData MakeFresh()
...
-                GetMergeResult(DiskIOInKB, other.DiskIOInKB));
+            var durationResult = GetMergeResult(DurationInMs, other.DurationInMs);
+            var peakMemoryResult = GetMergeResult(PeakMemoryInKB, other.PeakMemoryInKB);
+            var processorInPercentResult = GetMergeResult(ProcessorsInPercents, other.ProcessorsInPercents);
+            var diskIOResult = GetMergeResult(DiskIOInKB, other.DiskIOInKB);
+
+            return new PipHistoricPerfData(DefaultTimeToLive, durationResult, peakMemoryResult, (ushort)processorInPercentResult, diskIOResult);
         }
 
         private static uint GetMergeResult(uint oldData, uint newData)
```
Comment (mpysson) [inline Public/Src/Engine/Scheduler/PipHistoricPerfData.cs]: >GetMergeResult [](start = 28, length = 14)

How about making this internal and hitting it with some unit tests to bounds test the parameters?

File: Public/Src/Engine/Scheduler/PipHistoricPerfData.cs [hunk 3]
```diff
@@ -176,23 +177,30 @@ public PipHistoricPerfData MakeFresh()
...
-                GetMergeResult(DurationInMs, other.DurationInMs),
-                GetMergeResult(PeakMemoryInKB, other.PeakMemoryInKB),
-                (ushort)GetMergeResult(ProcessorsInPercents, other.ProcessorsInPercents),
-                GetMergeResult(DiskIOInKB, other.DiskIOInKB));
+            var durationResult = GetMergeResult(DurationInMs, other.DurationInMs);
+            var peakMemoryResult = GetMergeResult(PeakMemoryInKB, other.PeakMemoryInKB);
+            var processorInPercentResult = GetMergeResult(ProcessorsInPercents, other.ProcessorsInPercents);
+            var diskIOResult = GetMergeResult(DiskIOInKB, other.DiskIOInKB);
+
+            return new PipHistoricPerfData(DefaultTimeToLive, durationResult, peakMemoryResult, (ushort)processorInPercentResult, diskIOResult);
```
Comment (almiliMSFT) [inline Public/Src/Engine/Scheduler/PipHistoricPerfData.cs]: What's the change here?  Only reformatting?

File: Public/Src/Engine/Scheduler/PipHistoricPerfData.cs [hunk 4]
```diff
@@ -176,23 +177,30 @@ public PipHistoricPerfData MakeFresh()
...
-            // An asymmetric merge that goes up fast but decreases slowly.
-            if (newData > oldData)
+            try
             {
-                return (uint)(newData + (ulong) oldData) / 2;
-            }
+                // An asymmetric merge that goes up fast but decreases slowly.
+                if (newData > oldData)
+                {
+                    return (uint)(newData + (ulong) oldData) / 2;
```
Comment (almiliMSFT) [inline Public/Src/Engine/Scheduler/PipHistoricPerfData.cs]: how about changing to `newData/2 + oldData/2`?

File: Public/Src/Engine/Scheduler/PipHistoricPerfData.cs [hunk 5]
```diff
@@ -176,23 +177,30 @@ public PipHistoricPerfData MakeFresh()
...
+                }
 
-            return (uint)((((ulong)oldData * 9) + ((ulong)newData * 1)) / 10);
+                return (uint)((((ulong)oldData * 9) + ((ulong)newData * 1)) / 10);
+            }
+            catch(System.OverflowException ex)
```
Comment (almiliMSFT) [inline Public/Src/Engine/Scheduler/PipHistoricPerfData.cs]:
```suggestion
            catch (System.OverflowException ex)
```

File: Public/Src/Engine/Scheduler/PipHistoricPerfData.cs (hunk 3 above)
Comment (Kr1sso) [reply Public/Src/Engine/Scheduler/PipHistoricPerfData.cs]: Sort of, before the stack trace could not indicate any line number hence its was not easy to say which counter caused a potential overflow.

File: Public/Src/Engine/Scheduler/PipHistoricPerfData.cs (hunk 4 above)
Comment (Kr1sso) [reply Public/Src/Engine/Scheduler/PipHistoricPerfData.cs]: Indeed.

File: Public/Src/Engine/Processes/SandboxedProcessMac.cs (hunk 1 above)
Comment (Kr1sso) [reply Public/Src/Engine/Processes/SandboxedProcessMac.cs]: Thats the point exactly, merely casting does not throw an exception. This way we can at least get a better idea of where bogus data is coming from.

**Pull request merged**